torch/csrc/api/src/data/samplers/distributed.cpp

```diff
@@ -0,0 +1,164 @@
+#include <torch/data/samplers/distributed.h>
+#include <torch/serialize/archive.h>
+#include <torch/types.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <random>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+DistributedRandomSampler::DistributedRandomSampler(
+    size_t size,
+    size_t num_replicas,
+    size_t rank,
+    bool allow_duplicates)
+    : DistributedSampler(size, num_replicas, rank, allow_duplicates),
+      begin_index_(0),
+      end_index_(0),
+      sample_index_(0) {
+  populate_indices();
+  // shuffle first time.
+  reset(size_);
+}
+
+optional<std::vector<size_t>> DistributedRandomSampler::next(
+    size_t batch_size) {
+  if (sample_index_ == end_index_) {
+    return nullopt;
+  }
+
+  size_t end = sample_index_ + batch_size;
+  if (end > end_index_) {
+    end = end_index_;
+  }
+
+  auto iter = all_indices_.begin();
+  std::vector<size_t> res(iter + sample_index_, iter + end);
+  sample_index_ = end;
+  return res;
+}
+
+void DistributedRandomSampler::reset(optional<size_t> new_size) {
+  size_ = new_size.value_or(size_);  
+  populate_indices();
+  
+  std::mt19937 rand(epoch_);
+  std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
+  sample_index_ = begin_index_;
+}
+
+void DistributedRandomSampler::populate_indices() {
+  size_t sample_count =
+      num_replicas_ == 1 ? size_ : local_sample_count() * num_replicas_;
+  all_indices_.resize(sample_count);
+  std::iota(std::begin(all_indices_), std::end(all_indices_), 0);
+  for (size_t i = size_; i < sample_count; ++i) {
+    // we may have added duplicate samples to make all
+    // replicas to have the same number of samples.
+    all_indices_[i] = i % size_;
+  }
+  begin_index_ = rank_ * local_sample_count();
+  end_index_ = begin_index_ + local_sample_count();
+  sample_index_ = begin_index_;
+}
+
+void DistributedRandomSampler::save(serialize::OutputArchive& archive) const {
+  archive.write(
+      "sample_index_",
+      torch::tensor(static_cast<int64_t>(sample_index_)),
+      /*is_buffer=*/true);
+  archive.write(
+      "epoch_",
+      torch::tensor(static_cast<int64_t>(epoch_)),
+      /*is_buffer=*/true);
+}
+
+void DistributedRandomSampler::load(serialize::InputArchive& archive) {
+  auto tensor = torch::empty(1, torch::kInt64);
+  archive.read("sample_index_", tensor, /*is_buffer=*/true);
+  sample_index_ = tensor.item<int64_t>();
+
+  tensor = torch::empty(1, torch::kInt64);
+  archive.read("epoch_", tensor, /*is_buffer=*/true);
+  epoch_ = tensor.item<int64_t>();
```

Comment:
Don't you have to populate indices and then set the sample index? This doesn't restore the same state...

Reply:
I was under the impression that a reset will be called at the checkpoint, but that reset the index. The fix would be to call `populate_indices` after loading the size, index values.